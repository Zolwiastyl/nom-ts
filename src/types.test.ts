import { Brand, Flavor, Unbrand, Dict } from './index';

type PlayerId = Brand<number, 'PlayerId'>;
const PlayerId = (x: number) => x as PlayerId;

// @dts-jest:group (Brand)
{
  // @dts-jest:fail strings are assignable to branded numbers
  const _badId1: PlayerId = 'foobar';

  // @dts-jest:fail:snap internal type isn't assignable to branded type
  const _badId2: PlayerId = 1000;

  // @dts-jest:pass branded type is assignable to it's internal type
  const _goodId: number = PlayerId(1000);

  // @dts-jest:pass internal type prototype methods are accessible
  const _localeString = PlayerId(10000).toLocaleString(); //=> string
}

// @dts-jest:pass
const myId: PlayerId = PlayerId(1337);

type PlayerScores = Record<PlayerId, number>;

const scores: PlayerScores = { [myId]: 9001 };

// @dts-jest:fail:snap we can't index with branded type
const _1 = scores[myId];

type BetterPlayerScores = Partial<Dict<PlayerId, 9001 | 42 | 500>>;

const rf: BetterPlayerScores = {
  [1000]: 9001,
  [2000]: 42,
  [3000]: 500,
};

const id = PlayerId(1337);

// @dts-jest:pass we can extract internal type from Brand
const a = rf[Unbrand(id)]; //=> 9001 | 42 | 500 | undefined

type K = 'a' | 'b' | 'c';
type BK = Brand<K, 'BK'>;

// @dts-jest:pass
type DBK = Dict<BK, 1 | 2 | 3>;

const dbk: DBK = { a: 1, b: 2, c: 3 };

const bk = 'a' as BK;

// @dts-jest:pass
const _2 = dbk[Unbrand(bk)]; //=> 1 | 2 | 3

// @dts-jest:group (Flavor)
{
  type Flavored1 = Flavor<number, '1'>;
  const Flavored1 = (x: number) => x as Flavored1;

  type Flavored2 = Flavor<number, '2'>;

  // @dts-jest:pass internal type is assignable to flavor
  const _g1: Flavored1 = 1000;

  // @dts-jest:fail:snap one flavored type is not assignable to the other
  const _g2: Flavored2 = Flavored1(1000);

  // @dts-jest:pass internal type prototype methods are accessible
  const _fixed = Flavored1(10000).toFixed(); //=> string
}
