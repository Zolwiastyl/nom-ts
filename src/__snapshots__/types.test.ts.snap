// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Brand) internal type isn't assignable to branded type (type) should match snapshot: (Brand) internal type isn't assignable to branded type 1`] = `
"Type '1000' is not assignable to type 'Brand<number, \\"PlayerId\\">'.
  Type '1000' is not assignable to type '__Brand<number, \\"PlayerId\\">'."
`;

exports[`(Brand) we can't index with branded type (type) should match snapshot: (Brand) we can't index with branded type 1`] = `"Element implicitly has an 'any' type because expression of type 'Brand<number, \\"PlayerId\\">' can't be used to index type 'Record<Brand<number, \\"PlayerId\\">, number>'."`;

exports[`(Flavor) one flavored type is not assignable to the other (type) should match snapshot: (Flavor) one flavored type is not assignable to the other 1`] = `
"Type 'Flavor<number, \\"1\\">' is not assignable to type 'Flavor<number, \\"2\\">'.
  Type 'Flavor<number, \\"1\\">' is not assignable to type '__Flavor<number, \\"2\\">'.
    Types of property '[__brand]' are incompatible.
      Type '\\"1\\" | undefined' is not assignable to type '\\"2\\" | undefined'.
        Type '\\"1\\"' is not assignable to type '\\"2\\" | undefined'."
`;
